### PR TITLE
fix(core): s3-deployments don't work with new bootstrap stack

### LINKED ISSUE #8541: S3 Bucket Deployment does not work with new asset system
The S3 Bucket Deployment construct does not work anymore as the custom lambda does not have permission to use the KMS key of the asset bucket.

### Reproduction Steps

Use S3 Bucket Deployment

### Error Log

AccessDenied inside Deployment Lambda function

### Environment

  - **CLI Version      :** 1.45.0
  - **Framework Version:** 1.45.0
  - **Node.js Version:** v14
  - **OS               :** macOS
  - **Language (Version):** TypeScript

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by rix0rrr**
We added a KMS key, so the roles that want to read from the asset bucket should have KMS permissions.

Using `aws-s3-deployments` in combination with the new bootstrap stack requires a bootstrap stack update. You will see the following message if that is the case: `Stack failed: Error: Stack: publishing assets requires bootstrap stack version '3', found '2'. Please run 'cdk bootstrap' with a newer CLI version.`

Fixes #8541.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/core/lib/assets.ts [hunk 1]
```diff
@@ -208,6 +208,22 @@ export interface FileAssetLocation {
    * @example s3://mybucket/myobject
    */
   readonly s3ObjectUrl: string;
+
+  /**
+   * Whether the asset location has been encrypted by a KMS key
+   *
+   * If so, the consuming role should be given "kms:Decrypt" permissions in its
+   * identity policy. The reader role must get permissions to use all keys,
+   * since it's not possible to physical-name a key, and so we can't write the
```
Comment (eladb) [inline packages/@aws-cdk/core/lib/assets.ts:217]: I think @skinny85 added an alias

File: packages/@aws-cdk/core/lib/stack-synthesizers/default-synthesizer.ts [hunk 2]
```diff
@@ -247,6 +247,7 @@ export class DefaultStackSynthesizer implements IStackSynthesizer {
       httpUrl,
       s3ObjectUrl,
       s3Url: httpUrl,
+      kmsEncrypted: true,
```
Comment (eladb) [inline packages/@aws-cdk/core/lib/stack-synthesizers/default-synthesizer.ts:250]: The key alias should be propagated from here

**Comment by rix0rrr** [comment]
> I am not comfortable with granting decrypt permissions with any key. Feels unnecessarily promiscuous. I believe we should have an alias with a well known name in the bootstrap stack. Can we only grant permissions to it?

We don't have an alias.

We can add an alias, but it wouldn't help. It's not possible to add permissions to an alias, it must be on the key ID (which is a UUID).

**Comment by hoegertn** [comment]
What about a CFN Export of the key ID?

**Comment by rix0rrr** [comment]
> What about a CFN Export of the key ID?

Yep, good call. We ended up doing that.

File: packages/@aws-cdk/core/lib/stack-synthesizers/default-synthesizer.ts [hunk 3]
```diff
@@ -247,6 +267,7 @@ export class DefaultStackSynthesizer implements IStackSynthesizer {
       httpUrl,
       s3ObjectUrl,
       s3Url: httpUrl,
+      kmsKeyArn: Fn.importValue(cfnify(this._kmsKeyArnExportName!)),
```
Comment (eladb) [inline packages/@aws-cdk/core/lib/stack-synthesizers/default-synthesizer.ts:270]: curious, why do we need `!` here and in for `this.bucketName`? Generally I think we should completely avoid `!`. It's better to write the code that checks for nullity and throws an "unexpected" exception (or some other meaningful message) then let the null propagate to an unexpected place.

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.yaml [hunk 4]
```diff
@@ -347,14 +347,24 @@ Outputs:
     Description: The domain name of the S3 bucket owned by the CDK toolkit stack
     Value:
       Fn::Sub: "${StagingBucket.RegionalDomainName}"
+  FileAssetKeyArn:
```
Comment (eladb) [inline packages/aws-cdk/lib/api/bootstrap/bootstrap-template.yaml:350]: Do we want to add this to some integ test?

Comment (rix0rrr) [reply packages/aws-cdk/lib/api/bootstrap/bootstrap-template.yaml:350]: How about we add unit tests on the narrow contract of the bootstrap stack?

File: packages/@aws-cdk/aws-s3-assets/lib/asset.ts [hunk 5]
```diff
@@ -148,9 +146,13 @@ export class Asset extends cdk.Construct implements cdk.IAsset {
     this.s3ObjectUrl = location.s3ObjectUrl;
     this.httpUrl = location.httpUrl;
     this.s3Url = location.httpUrl; // for backwards compatibility
-    this.kmsKey = location.kmsKeyArn ? kms.Key.fromKeyArn(scope, 'Key', location.kmsKeyArn) : undefined;
 
-    this.bucket = s3.Bucket.fromBucketName(this, 'AssetBucket', this.s3BucketName);
+    const kmsKey = location.kmsKeyArn ? kms.Key.fromKeyArn(scope, 'Key', location.kmsKeyArn) : undefined;
```
Comment (eladb) [inline packages/@aws-cdk/aws-s3-assets/lib/asset.ts:150]: Do not pollute the parent scope:

```suggestion
    const kmsKey = location.kmsKeyArn ? kms.Key.fromKeyArn(this, 'Key', location.kmsKeyArn) : undefined;
```

File: packages/@aws-cdk/aws-s3-assets/test/asset.test.ts [hunk 6]
```diff
@@ -124,6 +124,31 @@ test('"readers" or "grantRead" can be used to grant read permissions on the asse
   });
 });
```
Comment (eladb) [inline packages/@aws-cdk/aws-s3-assets/test/asset.test.ts:126]: add a test to verify `asset.bucket` is properly configured (with the key)

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.yaml (hunk 4 above)
Comment (eladb) [reply packages/aws-cdk/lib/api/bootstrap/bootstrap-template.yaml:350]: Sounds good.. Anything that will protect us from changing it accidentally.

File: packages/@aws-cdk/aws-s3-assets/test/asset.test.ts (hunk 6 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-s3-assets/test/asset.test.ts:126]: Why?

The observable behavior (that a `grantRead()` leads to KMS permissions being added) is tested. That's all we need to care about?

File: packages/@aws-cdk/aws-s3-assets/lib/asset.ts (hunk 5 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-s3-assets/lib/asset.ts]: Ah right, good catch. Sorry about that.

File: packages/@aws-cdk/aws-s3-assets/test/asset.test.ts (hunk 6 above)
Comment (eladb) [reply packages/@aws-cdk/aws-s3-assets/test/asset.test.ts:126]: Got it. Thanks!

**Pull request merged**